sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -1196,7 +1196,7 @@ public static <DestinationT> FileResultCoder<DestinationT> of(
 
     @Override
     public List<? extends Coder<?>> getCoderArguments() {
-      return Arrays.asList(windowCoder);
+      return Arrays.asList(windowCoder, destinationCoder);
```

Comment:
Should it be FileResult<BoundedWindowT, DestinationT> and the windowCoder being Coder<BoundedWindowT extends BoundedWindow>? Looking from the code it think the window coder can change.

Reply:
You can see how it is used here: https://github.com/apache/beam/blob/243128a8fc52798e1b58b0cf1a271d95ee7aa241/sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java#L723

So if you have a `Coder<Foo<A, B, C>>` then the list of type arguments is `A, B, C` and `getCoderArguments` needs to return a `Coder<A>, Coder<B>, Coder<C>` that matches. So definitely before this pull request the coder registry will try to encode `DestinationT` with the window coder.

Right now `FileResult` just holds the window without tracking the type variable for it. I think it would be a breaking change to make it `FileResult<WindowT, DestinationT>`. I think just changing the window coder to the destination coder is the right thing to do.

Reply:
When registering a StructuredCoder, the component coders are hashed and compared in order to avoid duplicate. That means if we have one FileResultCoder(IntervalWindowCoder, VoidCoder) and another FileResultCoder(GlobalWIndowCoder, VoidCoder) they'll still cause collision.

I didn't see the window coder is being used to encode/decode the destinationT, the first registered FileResultCoder has serialized payload and then deserialized correctly and was able to decode the FileResult type correctly. But the other FileResultCoder will be deserialized from the same payload will throw exception because the destination coder does not match.

Reply:
getTypeToCoderBindings seems to be used when we need to infer coder from parameterized type? but it looks like we always call setCoder(FileResultCoder) at PCollections with FileResult type.

Reply:
Yes, to be clear, this is only used for automatically inferring coder from input to output. So like if you have `FileResult<DestinationT>` and you map it to `DestinationT` then just for that transform the coder will be automatically determined. Or at least that is the intent.

Reply:
So IIUC, the right thing to do is to have FileResult<WindowT, DestinationT> so that coder inference can work for both window type and destination type. But since all FileResult PCollections use explicit setCoders that's not a requirement.  I still think window coder needs to be in the components so that FileResultCoder with different window coders won't collide with each other.

Reply:
Yes, I think what you are describing makes sense. I do not know why we decided to not have a generic parameter for `<WindowT>`. Now it would be backwards-incompatible.

Note that `getComponents` is part of `StructuredCoder`. They overlap but are different. The things in `getComponents` should be any coder that can be passed to the constructor to change the behavior of the coder. Often there is no type variable.

So based on https://github.com/apache/beam/blob/e95ef975a3cb2f2562493ff2e0943d3e000376dc/sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java#L1187 the return value of `getComponents()` should be `windowCoder, destinationCoder` but because of https://github.com/apache/beam/blob/e95ef975a3cb2f2562493ff2e0943d3e000376dc/sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java#L1178 the return value of `getCoderArguments` should be just `destinationCoder`.

This class is actually a perfect example of why those two methods are different!

Reply:
Ah, that makes sense, update the PR.

Reply:
On a side note, I don't feel like the default behavior of getComponents() of StructuredCoder should just return whatever getCoderArguments() provides, this feels error-prone. Should we make getComponents a abstract method?